core/MyProtocol.cpp

```diff
@@ -111,14 +111,30 @@ bool protocolMQTT2MyMessage(MyMessage &message, char *topic, uint8_t *payload,
                             const unsigned int length)
 {
 	char *str, *p;
-	uint8_t index = 0;
+	int8_t index = -1;
 	message.setSender(GATEWAY_ADDRESS);
 	message.setLast(GATEWAY_ADDRESS);
 	message.setEcho(false);
-	for (str = strtok_r(topic + strlen(MY_MQTT_SUBSCRIBE_TOPIC_PREFIX) + 1, "/", &p);
+
+	// The subscription prefix can contain wildcards, but only the + wildcard.
+	// e.g. subscription prefix "+/mysensors", actual prefix "foo/mysensors"
+	// To fetch only the part after the mqtt prefix, discard as many
+	// mqtt levels (between / characters) as in the subscription prefix.
+	// Make a copy of the subscribe prefix because strtok is allowed to change the string.
+	char subscribeTopicPrefix[strlen(MY_MQTT_SUBSCRIBE_TOPIC_PREFIX) + 1];
+	strcpy(subscribeTopicPrefix,MY_MQTT_SUBSCRIBE_TOPIC_PREFIX);
+	char *strPrefix, *savePointerPrefix;
```

Comment:
What's the increase in ram/flash for this change?

Reply:
The savepointer (4 bytes?) is stored on stack I guess.
The increase in ram (stack use?) for the copy of the prefix string will be the length of the prefix plus 1.
The default value for the prefix is mygateway1-in, length is 13, so it will use 14 bytes on the stack. Your comment made me think about this memory usage. Actually I think malloc/free would be safer. Or maybe choose in runtime between stack and malloc depending on the prefix length? The length of the string is known at compile time, but I don't think that will help. I am not an expert in this subject.

Flash use? Is that the code length? The answer is: I don't know.

I compiled and ran it only on a raspberry pi 2, not on an Arduino.

Reply:
In core/MyGatewayTransportMQTTClient.cpp, method reconnectMQTT, there is also a local variable inTopic with a length of the prefix plus 10. So I think it is acceptable to have it as a stack variable.